web/src/pages/FilesPage.tsx

```diff
@@ -59,21 +58,61 @@ const FileRow = ({
 };
 
 export const FilesPage: React.FC = () => {
+  // The dataflow in this component is as follows:
+  // 1. The route changes which triggers the "loading" state
+  // 2. The current directory and its entities are fetched
+  // 3. If everything is okay, the "loading" state is removed
+  // to navigate to another entity, ONLY change the path through the history object
   const libraries = useService(LibrariesService);
   const { library: libraryId, path: encodedPath } = useParams<{ library: string; path?: string }>();
   const path = decodeURIComponent(encodedPath || '');
   const history = useHistory();
 
-  // Current directory info and details.
-  const [entries, setEntries] = useState<Entry[]>();
   const [currentDir, setCurrentDir] = useState<Entry>();
+  const [entries, setEntries] = useState<Entry[]>([]);
+  const [loadingEntries, setLoadingEntries] = useState(false);
+  const [shouldRefresh, setShouldRefresh] = useState(false);
+
+  // refresh when paths change
+  useEffect(() => {
+    setShouldRefresh(true);
+  }, [libraryId, path]);
+
+  // This effect triggers when the path changes. This means we should enter a loading state
+  useEffect(() => {
+    if (shouldRefresh) {
+      let mounted = true;
+      setLoadingEntries(true);
+      (async () => {
+        // If the path is falsy or '/' we're at the library root, so no need for an extra call.
+        const pathIsRoot = !path || path === '/';
+        // noinspection ES6MissingAwait It is awaited later to run in parallel
+        const getCurrentEntity = pathIsRoot ? Promise.resolve(undefined) : libraries.getEntry(libraryId, path);
+        const getCurrentChildren = libraries.getEntries(libraryId, path);
+
+        // TODO: Handle 404 on getCurrentEntity
+        const newCurrentDir = await getCurrentEntity;
+        const newEntries = await getCurrentChildren;
+        if (mounted) {
+          setCurrentDir(newCurrentDir);
+          setEntries(newEntries);
+          setLoadingEntries(false);
+          setShouldRefresh(false);
+        }
+      })();
+      return () => {
+        mounted = false;
+      };
+    }
+  }, [libraryId, libraries, path, shouldRefresh]);
+
+  // Current directory info and details.
   const onClickEntry = useCallback(
     async (target: string | Entry) => {
       if (typeof target === 'string') {
         history.push(`/files/${libraryId}/${encodeURIComponent(target)}`);
       } else if (target.category === 'Folder') {
         history.push(`/files/${libraryId}/${encodeURIComponent(resolvePath(target))}`);
-        setCurrentDir(target);
```

Comment:
This has the same problem as before: entering a folder will now briefly flash the current (to-be-previous) directory info in the details panel while it's loading the new current dir info - or just an empty panel (same goes for going up a directory btw). While I get this is the nice way of doing it in terms of dataflow, in terms of UX it's kinda bad imo.

![Peek 2021-07-24 14-10](https://user-images.githubusercontent.com/5592716/126868029-af6c24bb-d867-4299-a33c-ac3a0f8e579c.gif)

Reply:
Hm okay, I mapped them out. There were so many different issues with this way. I will think of a different more consistent way.

What would you expect? To have to current details panel there until it updates?

Reply:
Yeah that sounds like it would be good. That way when you navigate into a folder nothing should change, and navigating up will have a slight delay before changing, but that's fine.

Out of curiosity, what would be the issue with keeping the `setCurrentDir` call there? I don't see it in any effect dependencies, so it shouldn't trigger any unnecessary rerenders etc. Admittedly I haven't looked into the changes in much depth yet, but I fail to see the problem. (Not saying that if you have a nicer solution you shouldn't implement that of course)